tests/conftest.py

```diff
@@ -101,61 +100,87 @@ def chdir():
+    return d
+
+
 @pytest.fixture
-def data_archive(request, tmp_path_factory, chdir):
+def data_archive(request, tmp_path_factory):
     """
     Extract a .tgz data archive to a temporary folder.
 
     When --preserve-data is passed on the command line, a failing test will have the folder kept.
 
     Context-manager produces the directory path and sets the current working directory.
     """
-    incr = 0
 
     @contextlib.contextmanager
-    def _data_archive(name):
-        nonlocal incr
-
-        extract_dir = tmp_path_factory.mktemp(request.node.name, str(incr))
-        incr += 1
+    def ctx(name):
+        extract_dir = tmp_path_factory.mktemp(request.node.name, numbered=True)
         cleanup = True
         try:
-            archive_name = f"{name}.tgz"
-            archive_path = Path(__file__).parent / "data" / archive_name
-            with tarfile.open(archive_path) as archive:
-                archive.extractall(extract_dir)
-
-            L.info("Extracted %s to %s", archive_name, extract_dir)
-
-            # data archive should have a single dir at the top-level, matching the archive name.
-            assert (
-                len(os.listdir(extract_dir)) == 1
-            ), f"Expected {name}/ as the only top-level item in {archive_name}"
-            d = extract_dir / name
-            assert (
-                d.is_dir()
-            ), f"Expected {name}/ as the only top-level item in {archive_name}"
-
-            with chdir(d):
+            d = extract_archive(name, extract_dir)
+            with chdir_(d):
                 try:
                     yield d
-                except Exception:
+                finally:
                     if request.config.getoption("--preserve-data"):
                         L.info(
                             "Not cleaning up %s because --preserve-data was specified",
                             extract_dir,
                         )
                         cleanup = False
-                    raise
         finally:
             if cleanup:
                 time.sleep(1)
-                try:
-                    shutil.rmtree(extract_dir)
-                except PermissionError as e:
-                    L.debug("W: Issue cleaning up temporary folder: %s", e)
+                cleanup_dir(extract_dir)
+
+    return ctx
+
+
+@pytest.fixture()
+def data_archive_readonly(request, pytestconfig):
+    """
+    Extract a .tgz data archive to a temporary folder, and CACHE it in the pytest cache.
+    Don't use this if you're going to write to the dir!
+    If you don't need to write to the extracted archive, use this in preference to data_archive.
+
+    Context-manager produces the directory path and sets the current working directory.
+    """
+
+    @contextlib.contextmanager
+    def ctx(name):
+        root = Path(request.config.cache.makedir('data_archive_readonly'))
+        path = root / name
+        if path.exists():
+            L.info("Found cache at %s", path)
```

Comment:
How does caching work — if we update the tgz does it automatically flush somehow? Or do we do it based off a hash/something of the archive?

Reply:
At present you have to clear the cache via `pytest --cache-clear ...`

We'd have to read the whole thing to hash it. Presumably that's still (much) faster than extracting it, but it's not ideal...

But maybe it's worth it to avoid confusion if/when the archives change

Reply:
Done in 15dbaa8b927a171e13c7b44ed29d7e6d869a8de3